### DISCUSSION TITLE
<Solved> Will not bind to webgui IP at startup, but starts fine when manually started. <Solved>

### DISCUSSION THREAD

**Discussion by Huecuva** [jcorporation/myMPD#510] [Q&A] [answered]
Hey guys. I have myMPD set up and working fine on a SFF PC I have using bare metal and it's working just fine. However, my actual server is a Debian VM running in Proxmox. I have my mpd stream set up and working, and I have myMPD mostly working, except that when my VM starts, the myMPD service fails to bind to the configured web interface IP address. The weird part is, if I ssh into the server and manually start the myMPD service with `systemctl start mympd` it starts without an issue and I'm able to access it via a browser on my local machine. 

The errors I'm getting in `systemctl status mympd` and `journalctl | grep mympd` are thus:
```
# systemctl status mympd
● mympd.service - myMPD server daemon
   Loaded: loaded (/lib/systemd/system/mympd.service; enabled; vendor preset: enabled)
   Active: failed (Result: exit-code) since Wed 2021-06-09 20:26:10 PDT; 30min ago
  Process: 545 ExecStart=/usr/bin/mympd (code=exited, status=1/FAILURE)
 Main PID: 545 (code=exited, status=1/FAILURE)

Jun 09 20:26:10 Ilithid mympd[545]: NOTICE   mympd     Setting loglevel to NOTICE
Jun 09 20:26:10 Ilithid mympd[545]: NOTICE   mympd     Localstate dir: "/var/lib/mympd"
Jun 09 20:26:10 Ilithid mympd[545]: NOTICE   mympd     SSL certificates: "/var/lib/mympd/ssl"
Jun 09 20:26:10 Ilithid mympd[545]: NOTICE   mympd     CA certificate and private key found
Jun 09 20:26:10 Ilithid mympd[545]: NOTICE   mympd     Server certificate and private key found
Jun 09 20:26:10 Ilithid mympd[545]: 2021-06-10 03:26:10  E mongoose.c:2549:mg_open_l Failed to listen on http://10.192.10.149:9999, errno 99
Jun 09 20:26:10 Ilithid mympd[545]: ERROR    mympd     Can't bind to http://10.192.10.149:9999
Jun 09 20:26:10 Ilithid mympd[545]: 2021-06-10 03:26:10  I mongoose.c:2027:mg_mgr_fr All connections closed
Jun 09 20:26:10 Ilithid systemd[1]: mympd.service: Main process exited, code=exited, status=1/FAILURE
Jun 09 20:26:10 Ilithid systemd[1]: mympd.service: Failed with result 'exit-code'.
# journalctl | grep mympd
Jun 09 20:26:10 Ilithid mympd[545]: NOTICE   mympd     Starting myMPD 7.0.2
Jun 09 20:26:10 Ilithid mympd[545]: NOTICE   mympd     Libmympdclient 1.0.5 based on libmpdclient 2.20.0
Jun 09 20:26:10 Ilithid mympd[545]: NOTICE   mympd     Mongoose 7.3
Jun 09 20:26:10 Ilithid mympd[545]: NOTICE   mympd     Parsing config file: /etc/mympd.conf
Jun 09 20:26:10 Ilithid mympd[545]: NOTICE   mympd     Scripting disabled, disabling scripteditor
Jun 09 20:26:10 Ilithid mympd[545]: NOTICE   mympd     Setting loglevel to NOTICE
Jun 09 20:26:10 Ilithid mympd[545]: NOTICE   mympd     Localstate dir: "/var/lib/mympd"
Jun 09 20:26:10 Ilithid mympd[545]: NOTICE   mympd     SSL certificates: "/var/lib/mympd/ssl"
Jun 09 20:26:10 Ilithid mympd[545]: NOTICE   mympd     CA certificate and private key found
Jun 09 20:26:10 Ilithid mympd[545]: NOTICE   mympd     Server certificate and private key found
Jun 09 20:26:10 Ilithid mympd[545]: 2021-06-10 03:26:10  E mongoose.c:2549:mg_open_l Failed to listen on http://10.192.10.149:9999, errno 99
Jun 09 20:26:10 Ilithid mympd[545]: ERROR    mympd     Can't bind to http://10.192.10.149:9999
Jun 09 20:26:10 Ilithid mympd[545]: 2021-06-10 03:26:10  I mongoose.c:2027:mg_mgr_fr All connections closed
Jun 09 20:26:10 Ilithid systemd[1]: mympd.service: Main process exited, code=exited, status=1/FAILURE
Jun 09 20:26:10 Ilithid systemd[1]: mympd.service: Failed with result 'exit-code'.
# systemctl start mympd
# systemctl status mympd
● mympd.service - myMPD server daemon
   Loaded: loaded (/lib/systemd/system/mympd.service; enabled; vendor preset: enabled)
   Active: active (running) since Wed 2021-06-09 21:18:34 PDT; 2s ago
 Main PID: 660 (mympd)
    Tasks: 5 (limit: 4915)
   Memory: 13.4M
   CGroup: /system.slice/mympd.service
           └─660 /usr/bin/mympd
```
Here are the relevant parts of /etc/mympd.conf:
```
[webserver]
#Webserver options
httphost = 10.192.10.149

httpport = 9999
```
You can see that it does in fact start when started manually. I don't know why this is happening and my /etc/mympd.conf file in the VM server is identical to the /etc/mympd.conf file in my SFF mpd server where relevant. Is there some kind of timeout issue? Does anyone have any ideas?

I could just set up a cron job to start the mympd service after a delay, but that seems hacky to me and I would prefer to entertain cleaner solutions until that becomes necessary.

Thanks.

**Comment by jcorporation** [comment, marked as answer]
This could be a timing issue of the boot process. Sure the network is up before the system starts myMPD in the boot process?

**Reply by Huecuva** [reply]
I'm not sure at all, but given that the mpd service starts without issue and the contents of /usr/lib/systemd/system/mpd.service: 
```
[Unit]
Description=Music Player Daemon
Documentation=man:mpd(1) man:mpd.conf(5)
Documentation=file:///usr/share/doc/mpd/user-manual.html
After=network.target sound.target

[Install]
WantedBy=multi-user.target
Also=mpd.socket
```

and /usr/lib/systemd/system/mympd.service:

```
[Unit]
Description=myMPD server daemon
Requires=network.target local-fs.target
After=mpd.service

[Install]
WantedBy=multi-user.target
```

It appears as if everything should be starting in the correct order. I'm not sure how to fix what doesn't appear to be broken.

**Reply by jcorporation** [reply]
This is often the case, finding the error is the bigest problem. Is there a boot log where you can investigate the boot order?

Have you tried setting httphost to 0.0.0.0?

**Reply by Huecuva** [reply]
I don't know how to find out in what order the services start. However, I just tried setting the httphost to 0.0.0.0 and now the mympd webgui binds properly and is accessible on boot, however the local playback doesn't work anymore and will no longer connect to the stream port.

EDIT: Scratch that. It appears that the local playback wasn't binding because I had stopped the mpd stream. Once I actually picked a track and clicked play, it was able to connect to the stream port and start playing. Setting the httphost to 0.0.0.0 seems to have solved my issue.

Thanks!
